### PR TITLE
Support pretty URLs

### LINKED ISSUE #5744: Get rid of index.php in the url
to be clear this is optional and we cannot rely on this.

more info to come ....

### LINKED ISSUE #12345: Document front_controller_active in config.sample.php
From https://github.com/owncloud/core/issues/8269#issuecomment-63934300

>    @DeepDiver1975 I checked the sample file on my 7.0.3 installation and https://github.com/owncloud/core/blob/master/config/config.sample.php in the repo. I couldn't find any reference to index.php.

Looks like the config flag is not documented - sorry for this. The config setting is named 'front_controller_active' and it should be set to true.

### PR DESCRIPTION
**Pull request by LukasReschke**
This changeset allows ownCloud to run with pretty URLs, they will be used by default if mod_rewrite and mod_env are available. This means basically that the `index.php` in the URL is not shown to the user anymore.

Also the not deprecated functions to generate URLs have been modified to support this behaviour, old functions such as `filePath` will still behave as before for compatibility reasons.

Examples:
http://localhost/index.php/s/AIDyKbxiRZWAAjP => http://localhost/s/AIDyKbxiRZWAAjP
http://localhost/index.php/apps/files/ => http://localhost/apps/files/

Due to the way our CSS and JS is structured the .htaccess uses some hacks for the final result but could be worse... And I was just annoyed by all that users crying for the removal of `index.php` ;-)

@karlitschek @DeepDiver1975 Please decide if this is something we want to have. If yes I'll see how we could best unit-test this…

Fixes https://github.com/owncloud/core/issues/5744

Fixes #12345

### CODE REVIEW

**Comment by LukasReschke** [comment]
This is obviously only for Apache as I have no idea at all how to use or configure other webservers ;-)

**Comment by LukasReschke** [comment]
To test enable mod_rewrite and mod_env, things should work magically. Old URLs with `index.php` will still work and not get redirected since I didn't want to have too many things exploding.

**Comment by LukasReschke** [comment]
Please notice that supporting this obviously also requires twice as much QA ;-)

(i.e. on a server with mod_rewrite installed and another one without)

File: .htaccess [hunk 1]
```diff
@@ -30,6 +34,18 @@ RewriteRule ^apps/contacts/carddav\.php remote.php/carddav/ [QSA,L]
 RewriteRule ^remote/(.*) remote.php [QSA,L]
 RewriteRule ^(build|tests|config|lib|3rdparty|templates)/.* - [R=404,L]
 RewriteRule ^(\.|autotest|occ|issue|indie|db_|console).* - [R=404,L]
+
+# Pretty URLs without leading /index.php/
+RewriteRule ^core/js/oc.js index.php/$0 [PT]
+RewriteRule ^core/preview.png index.php/$0 [PT]
+RewriteCond %{REQUEST_FILENAME} \.(css|js|svg|gif|png|html|ttf|woff)$ [OR]
```
Comment (LukasReschke) [inline .htaccess]: Black voodoo because there is no static folder (as in most frameworks such as Symfony) and thus we have to do that guessing here.

File: .htaccess [hunk 2]
```diff
@@ -30,6 +34,18 @@ RewriteRule ^apps/contacts/carddav\.php remote.php/carddav/ [QSA,L]
...
+RewriteCond %{REQUEST_FILENAME} /index.php [OR]
+RewriteCond %{REQUEST_FILENAME} /remote.php [OR]
+RewriteCond %{REQUEST_FILENAME} /public.php [OR]
+RewriteCond %{REQUEST_FILENAME} /status.php [OR]
+RewriteCond %{REQUEST_FILENAME} /ocs/v1.php
```
Comment (LukasReschke) [inline .htaccess]: Other black voodoo for all allowed endpoints. But at least it gives us some additional security as well for free as all other PHP files will not be directly accessible anymore :smile:

**Comment by oparoz** [comment]
Subscribing

**Comment by DeepDiver1975** [comment]
great initiative @LukasReschke THX

I already did introduce a config flag for removing index.php from the generated urls - maybe you can reuse this one - https://github.com/owncloud/core/blob/use-pretty-urls/lib/private/urlgenerator.php#L70

**Comment by LukasReschke** [comment]
> I already did introduce a config flag for removing index.php from the generated urls - maybe you can reuse this one - https://github.com/owncloud/core/blob/use-pretty-urls/lib/private/urlgenerator.php#L70

I think it would make more sense if we stick with the automatic approach implemented here instead of relying on some configuration switch. As far I understand most people don't change anything in the config and neither do we devs, so if something is standard in most environments it gets automatically better tested. But we can also make this a configuration only option if desired.

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.

**Comment by shoeper** [comment]
To follow this automatic approach we could add a rewrite rule to the nginx config in the manual and set an env to detect the redirect rule.
see http://stackoverflow.com/questions/8098927/nginx-variables-similar-to-setenv-in-apache
Something like
```
rewrite ...
env[REMOVE_INDEX_PHP] = true
```

**Comment by Xenopathic** [comment]
The nginx configuration is very simple, and in fact doesn't need any rewrites at all. I'm using:

```
try_files $uri $uri/ /index.php?$query_string;
location ~ \.php(?:$|/) {
    fastcgi_split_path_info ^(.+?\.php)(/.+)$;
    fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
    fastcgi_param PATH_INFO $fastcgi_path_info;
    fastcgi_pass ...;
}
```

However, getting that environment variable into PHP-FPM is proving difficult. I tried `fastcgi_param` in my nginx config and `env[modRewriteWorking] = true' in PHP-FPM's config, but neither worked. I know that with the above nginx config, pretty URLs work (I manually removed index.php from the URL bar), but that worked before this PR too.

I think @DeepDiver1975 's suggestion of using a config.php variable is the best. It makes it much easier for other web servers, and is actually easier to set that way. I know we lose the automatic aspect, but setting something in config.php is easy, while setting something in PHP-FPM or nginx.conf (and even then I don't know what) is more difficult.

**Comment by shoeper** [comment]
Setting the env in the config wouldn't be an issue if possible because most 
users will just copy the nginx config from owncloud manual.

But a config setting would be enough. In my opinion people not being able 
to just set a config option shouldn't run owncloud because they aren't able 
to assess risks and won't be able to securely setup their installation.


Am 11. Februar 2015 19:31:40 schrieb Robin McCorkell 
<notifications@github.com>:

**Comment by DeepDiver1975** [comment]
Guess what? Doesn't work for me :speak_no_evil: 

bildschirmfoto von 2015-02-11 20 39 47

**Comment by LukasReschke** [comment]
As discussed with @DeepDiver1975 we can use the automatic approach since this will lead to problems in some deployment scenarios and our goal is to be as easy to install as possible.

Thus I migrated this to the `front_controller_active` config setting which defaults to `false` and added some documentation in `config.sample.php` and `.htaccess`

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.

**Comment by LukasReschke** [comment]
Added some basic unit tests since this is not really testable the approach is somewhat hacky. Ready to review now.

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.

**Comment by DeepDiver1975** [comment]
tested :+1:

**Comment by LukasReschke** [comment]
Probably needs some magic unit tests adjustments - love that static thingies… On it…

**Comment by DeepDiver1975** [comment]
iqtce

**Comment by MorrisJobke** [comment]
@owncloud-bot Retest this please.

File: config/config.sample.php [hunk 3]
```diff
@@ -240,6 +240,18 @@
...
+ * with clean URLs. This means that the ``index.php`` gets removed from the URL
+ * visible to the user and the URL is therefore shorter and by some people
+ * considered cleaner.
+ *
+ * Setting this switch to ``true`` requires to have the the web server properly
+ * configured as well, ownCloud comes with an example configuration for Apache in
+ * combination with ``mod_rewrite`` those rules can be seen and enabled in the
+ * ``.htaccess`` file in the ownCloud root folder.
+ */
+'front_controller_active' => false,
```
Comment (MorrisJobke) [inline config/config.sample.php]: Keep in mind, that in the route generator `'false'` and `'true'` as string are used ;)

**Comment by MorrisJobke** [comment]
```
07:51:01 There were 2 failures:
07:51:01 
07:51:01 1) Test_Util::testGetDefaultPageUrlWithRedirectUrlWithoutFrontController
07:51:01 Failed asserting that two strings are identical.
07:51:01 --- Expected
07:51:01 +++ Actual
07:51:01 @@ @@
07:51:01 -http://localhost//owncloudmyRedirectUrl.com
07:51:01 +http://localhost/owncloud/myRedirectUrl.com
07:51:01 
07:51:01 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/tests/lib/util.php:411
07:51:01 
07:51:01 2) Test_Util::testGetDefaultPageUrlWithRedirectUrlRedirectBypassWithoutFrontController
07:51:01 Failed asserting that two strings are identical.
07:51:01 --- Expected
07:51:01 +++ Actual
07:51:01 @@ @@
07:51:01 -http://localhost//owncloudindex.php/apps/files/
07:51:01 +http://localhost/owncloud/index.php/apps/files/
07:51:01 
07:51:01 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/tests/lib/util.php:422
```

File: config/config.sample.php (hunk 3 above)
Comment (LukasReschke) [inline config/config.sample.php]: :see_no_evil: – let's change that as well – grrr…

Comment (MorrisJobke) [inline config/config.sample.php]: Then keep in mind to add a repair step for existing installations ;)

Comment (LukasReschke) [inline config/config.sample.php]: Why? It was undocumented and unused :)

Comment (LukasReschke) [inline config/config.sample.php]: (as in: "the config option didn't do anything at all before")

**Comment by Xenopathic** [comment]
I'll try and get an nginx config that lets this work, ASAP.

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

(and don't link to the wrong result set!!!)

**Comment by icewind1991** [comment]
Seems to work fine with php+nginx using 

```
    rewrite  ^/owncloud/(core\/(preview.png|js\/oc.js))$  /owncloud/index.php/$1  last;
    rewrite  ^/owncloud/([^.]*)$  /owncloud/index.php/$1  last;
```

**Comment by icewind1991** [comment]
:+1:

**Comment by MorrisJobke** [comment]
@PVince81 ;)

**Comment by stefarossi** [comment]
Just wondering, when do you think this feature will be released?

**Comment by DeepDiver1975** [comment]
@owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
@owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by oparoz** [comment]
Try to simply clone the branch in another PR to get the bot to test it. In one of my PRs, it even added the result back to the original one.

**Comment by MorrisJobke** [comment]
Jenkins PR #15408

**Comment by MorrisJobke** [comment]
Looks like broken URLs:

```
00:13:10.933 There were 5 failures:
00:13:10.933 
00:13:10.933 1) Test_Util::testFormatDateWithTZFromSession with data set #3 (9.5, 'October 13, 2012 at 9:23:25 P...T+9:30')
00:13:10.933 Failed asserting that two strings are equal.
00:13:10.933 --- Expected
00:13:10.933 +++ Actual
00:13:10.933 @@ @@
00:13:10.933 -'October 13, 2012 at 9:23:25 PM GMT+9:30'
00:13:10.933 +'October 13, 2012 at 10:23:25 PM GMT+10:30'
00:13:10.933 
00:13:10.933 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/util.php:78
00:13:10.933 
00:13:10.933 2) Test_Util::testGetDefaultPageUrlWithRedirectUrlWithoutFrontController
00:13:10.936 Failed asserting that two strings are identical.
00:13:10.936 --- Expected
00:13:10.936 +++ Actual
00:13:10.936 @@ @@
00:13:10.936 -http://localhost//owncloudmyRedirectUrl.com
00:13:10.936 +http://localhost/owncloud/myRedirectUrl.com
00:13:10.936 
00:13:10.936 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/util.php:401
00:13:10.936 
00:13:10.936 3) Test_Util::testGetDefaultPageUrlWithRedirectUrlRedirectBypassWithoutFrontController
00:13:10.937 Failed asserting that two strings are identical.
00:13:10.937 --- Expected
00:13:10.937 +++ Actual
00:13:10.937 @@ @@
00:13:10.937 -http://localhost//owncloudindex.php/apps/files/
00:13:10.937 +http://localhost/owncloud/index.php/apps/files/
00:13:10.937 
00:13:10.937 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/util.php:412
00:13:10.937 
00:13:10.937 4) OCA\Files\Controller\ApiControllerTest::testGetFilesByTagSingle
00:13:10.937 Failed asserting that two objects are equal.
00:13:10.937 --- Expected
00:13:10.937 +++ Actual
00:13:10.937 @@ @@
00:13:10.937                  'parentId' => null
00:13:10.937 -                'date' => 'January 1, 1970 at 12:00:55 AM GMT+0'
00:13:10.937 +                'date' => 'January 1, 1970 at 9:30:55 AM GMT+9:30'
00:13:10.937                  'mtime' => 55000
00:13:10.937                  'icon' => '/owncloud/core/img/filetypes/...df.svg'
00:13:10.937                  'name' => 'root.txt'
00:13:10.937                  'permissions' => null
00:13:10.937                  'mimetype' => 'application/pdf'
00:13:10.937                  'size' => 1234
00:13:10.937                  'type' => 'file'
00:13:10.937                  'etag' => 'MyEtag'
00:13:10.937                  'path' => '/'
00:13:10.937                  'tags' => Array (...)
00:13:10.937              )
00:13:10.937          )
00:13:10.937      )
00:13:10.937      'headers' => Array (...)
00:13:10.937      'cookies' => Array ()
00:13:10.937      'status' => 200
00:13:10.937      'lastModified' => null
00:13:10.937      'ETag' => null
00:13:10.937      'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
00:13:10.937  )
00:13:10.937 
00:13:10.937 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files/tests/controller/apicontrollertest.php:131
00:13:10.937 
00:13:10.937 5) OCA\Files\Controller\ApiControllerTest::testGetFilesByTagMultiple
00:13:10.938 Failed asserting that two objects are equal.
00:13:10.938 --- Expected
00:13:10.938 +++ Actual
00:13:10.938 @@ @@
00:13:10.938                  'parentId' => null
00:13:10.938 -                'date' => 'January 1, 1970 at 12:00:55 AM GMT+0'
00:13:10.938 +                'date' => 'January 1, 1970 at 9:30:55 AM GMT+9:30'
00:13:10.938 
00:13:10.938 @@ @@
00:13:10.938                  'parentId' => null
00:13:10.938 -                'date' => 'January 1, 1970 at 12:16:39 AM GMT+0'
00:13:10.938 +                'date' => 'January 1, 1970 at 9:46:39 AM GMT+9:30'
00:13:10.938                  'mtime' => 999000
00:13:10.938                  'icon' => '/owncloud/core/img/filetypes/...on.svg'
00:13:10.938                  'name' => 'root.txt'
00:13:10.938                  'permissions' => null
00:13:10.938                  'mimetype' => 'application/binary'
00:13:10.938                  'size' => 9876
00:13:10.938                  'type' => 'file'
00:13:10.938                  'etag' => 'SubEtag'
00:13:10.938                  'path' => '/'
00:13:10.938                  'tags' => Array (...)
00:13:10.938              )
00:13:10.938          )
00:13:10.938      )
00:13:10.938      'headers' => Array (...)
00:13:10.938      'cookies' => Array ()
00:13:10.938      'status' => 200
00:13:10.938      'lastModified' => null
00:13:10.938      'ETag' => null
00:13:10.938      'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
00:13:10.938  )
00:13:10.938 
00:13:10.938 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files/tests/controller/apicontrollertest.php:215
```

The time unit tests will hopefully be fixed by a rebase?!?

**Comment by MorrisJobke** [comment]
@DeepDiver1975 @LukasReschke Something for 8.2?

**Comment by DeepDiver1975** [comment]
8.2 yes

**Comment by msva** [comment]
btw, guys, don't you forget to also get rid of "remote.php" in urls too? :)

**Comment by msva** [comment]
Btw, that way of getting webroot collides with front_controller_active (and causes 404s): https://github.com/owncloud/core/blob/master/core/js/js.js#L18 .
// it returns "/apps/files/" if called from files app, while it shoud return real OC root path :-/

**Comment by messi89** [comment]
Hi guys

is there any solution to remove index.php from url ?

**Comment by LukasReschke** [comment]
> btw, guys, don't you forget to also get rid of "remote.php" and in DAV urls too? :)

What is the actual gain? These URIs are not aimed to be really displayed to users. Not worth to invest a lot of time in this thus. – I mean that ownCloud runs on PHP is already exposed since we're open-source :speak_no_evil: 

> is there any solution to remove index.php from url ?

Not yet. I shall revive this in time for 8.2

**Comment by msva** [comment]
> What is the actual gain? These URIs are not aimed to be really displayed to users. Not worth to invest a lot of time in this thus. – I mean that ownCloud runs on PHP is already exposed since we're open-source :speak_no_evil:

Original goal is to prettify (and shorten) {web,cal,card}DAV links in OC web-interface ;)

**Comment by LukasReschke** [comment]
> Original goal is to prettify (and shorten) {web,cal,card}DAV links in OC web-interface ;)

I won't do this. If you want to take care of this you're welcomr to create a PR based on this one once it is merged. - I'm really not keen on using mod_rewrite magic on any of our DAV backends

**Comment by DeepDiver1975** [comment]
at least caldav and carddav service lookup can we made easier by using .well-known or DNS service entries - http://sabre.io/dav/service-discovery/

**Comment by nanocosmos** [comment]
Any update on the original issue removing index.php? What is the easiest way to do this?

**Comment by MorrisJobke** [comment]
@LukasReschke Defer this to 9.0 or close it?

**Comment by hutchic** [comment]
rebasing this to master fixed the test failures for me

**Comment by MorrisJobke** [comment]
@LukasReschke rebased and conflicts resolved ;)

**Comment by mahnunchik** [comment]
I look forward to pretty urls :smile:

**Comment by DeepDiver1975** [comment]
rebased

@cburnard this is the change we have been chatting about with one of our customers these days

@karlitschek @MTRichards @cmonteroluque  FYI

@LukasReschke what is left to be done here? Please add list of TODOs - THX

**Comment by cburnard** [comment]
> @cburnard this is the change we have been chatting about with one of our customers these days

@DeepDiver1975 Thanks!

**Comment by LukasReschke** [comment]
@DeepDiver1975 I fixed the rewrite rules. All routes work for me locally now.
@rperezb Stuff for the test plan. Basically means that all functionalities have to be tested twice. Once with this setting enabled and once without.

**Comment by LukasReschke** [comment]
… another one security enhancement by this is that it blocks access to everything that cannot be accessed via the front controller. Pretty cool.

**Comment by MorrisJobke** [comment]
Looks pretty cool :)

**Comment by LukasReschke** [comment]
Rebased again.

**Comment by MorrisJobke** [comment]
* set front_controller_active to true
* uncommented the rewrite rules
* prepend "RewriteBase /master/" because my instance is in the subfolder master


-> doesn't work:

bildschirmfoto 2015-12-01 um 08 41 10

**Comment by LukasReschke** [comment]
@MorrisJobke Then you do something wrong. Let's screenshare later. Also RewriteBase is not required for subfolders ;)

File: lib/private/setup.php [hunk 4]
```diff
@@ -432,6 +432,12 @@ public static function updateHtaccess() {
 			//custom 404 error page
 			$content.= "\nErrorDocument 404 ".\OC::$WEBROOT."/core/templates/404.php";
 		}
+
+		// Add rewrite base
+		$content.="\n<IfModule mod_rewrite.c>";
+		$content.="\nRewriteBase ".\OC::$WEBROOT;
```
Comment (MorrisJobke) [inline lib/private/setup.php]: Can we add two spaces after the `\n` so the `RewriteBase` is intended? That would be nice :)

**Comment by MorrisJobke** [comment]
I retested and it still works. Now even without changing anything. It also works with existing instances and during a fresh install.

So here is my :+1: from me

**Comment by MorrisJobke** [comment]
Once this is in, we maybe should also bring in the NGINX gurus and have the NGINX doc updated ;)

File: lib/private/setup.php (hunk 4 above)
Comment (LukasReschke) [inline lib/private/setup.php]: Done.

Comment (MorrisJobke) [inline lib/private/setup.php]: :stuck_out_tongue_winking_eye:

**Comment by rullzer** [comment]
Awesome, everything still seems to work when clicking around!

:+1:

**Comment by mitar** [comment]
Something like that should probably be done in nginx;

```
location / {
    index index.php index.html index.htm;
    try_files $uri $uri/ /index.php/$uri?=$args;
}
````

**Comment by MorrisJobke** [comment]
cc @josh4trunks 

FYI: This only removes the index.php. remote.php and ocs/v1.php are still in the URL. ;)

**Comment by josh4trunks** [comment]
@mitar @MorrisJobke The documented nginx config already has the following
```
index index.php;
try_files $uri $uri/ /index.php;
```

I don't see a reason to include index.html/index.htm when these aren't present very often in owncloud, and as far I know owncloud doesn't use arguments so the try_files line we have is fine as is. Finally if you are going to use args with nginx the cleanest way is ``index.php$is_args$args``

**Comment by mitar** [comment]
The point is that the resulting URL is just internal, so `/index.php?$args` is good enough, no?

OwnCloud might not use them, but plugins might?

**Comment by mitar** [comment]
So there is no need to provide `$uri` to `index.php`?

**Comment by josh4trunks** [comment]
As I stated, "/index.php" is already in the documented nginx config. and try_files doesn't "provide" $uri to index.php, it tries the provided uri, uri with a slash (which would check "$uri/index.php"), then finally redirects to the main /index.php page.

But if apps do use arguments we should replace "/index.php" with "/index.php$is_args$args" in the try_files line in the documented config.

**Comment by stevenmirabito** [comment]
Here's my go at a working Nginx configuration: https://gist.github.com/stevenmirabito/5560ba2feea9eeb32db1

I've been ironing out the kinks and everything seems to be working on my installation. I did have a problem with the change to `core/js/js.js` however - POST requests without /index.php/ in the URL will fail as you can only rewrite GET or HEAD requests, and the requests need to be made through index.php for everything to load properly. I'm not sure how this is working on Apache with the given configuration, unless I'm doing something completely wrong in mine, but for now I've reverted that change to my installation.

Unfortunately, due to the way ownCloud is set up, a simple `try_files` statement won't work here - requesting `https://myowncloud.tld/apps/files/` will load that directory directly as it exists in the file structure instead of through index.php. As a result, I had to resort to messy regex whitelisting to reproduce what was accomplished in the committed .htaccess.

Let me know if anyone has any questions or suggestions. Hope this helps!

**Comment by josh4trunks** [comment]
@stevenmirabito what are you trying to accomplish? hiding index.php from the URL?
so a client can connect to /apps/files instead of /index.php/apps/files?

**Comment by josh4trunks** [comment]
@mitar I just realized what you were trying to say earlier, my mistake you were trying to pass the uri to index.php, not just arguments, (I missed that part!)

so @stevenmirabito you tried ``try_files $uri $uri/ /index.php$uri``

**Comment by stevenmirabito** [comment]
@josh4trunks Yes, that's the entire point of this pull request. @LukasReschke provided the Apache mod_rewrite config, what I posted above is for Nginx.

Like I said, I did try a `try_files` directive, but it won't work here because what that does is looks to see if the file actually exists, then passes it to index.php if it doesn't. For ownCloud, the files do exist, so that directive will not work (for example, requesting `/apps/files/` will prompt Nginx to try to load `/apps/files/index.php`, which it will as that file exists).

**Comment by guidtz** [comment]
Hello,
  I have an installation with Debian 8 / Apache 2.4 and Owncloud 8.2.2 but index.php stell apear in url.
There is a solution in this version or we have to wait 9.0 version ?

Thanks

**Comment by IngwiePhoenix** [comment]
I am using your VBox Appliance. How can I tell that to utilize pretty URLs?

**Comment by DeepDiver1975** [comment]
This is a New feature of 9.0

**Comment by RealRancor** [comment]
@josh4trunks @stevenmirabito @mitar Let's move the discussion about the needed changes in the nginx config to https://github.com/owncloud/documentation/issues/2014

Havn't noticed your posts in here and that info could be really useful there.

**Comment by thejustsoul** [comment]
> This is a New feature of 9.0
How to enable this for **Apache2**? To **automatically** (maybe 302 redirect?) used the URL:
https://localhost.tld/index.php/s/AIDyKbxiRZWAAjP => https://localhost.tld/s/AIDyKbxiRZWAAjP
https://localhost.tld/index.php/settings/admin => https://localhost.tld/settings/admin
etc..
And to "share" file, the link was without index.php in the address.

**Comment by pandorra2** [comment]
huh, did i get all of you wrong?
the only yet implemented solution ive found is this one, well documented in the 9.0 manual:
https://doc.owncloud.org/server/9.0/admin_manual/configuration_server/config_sample_php_parameters.html
(ctrl+f RewriteBase)

so i've added `'htaccess.RewriteBase' => '/',` to my config. well, doesn't work - right as the manual warns already. still have to run that occ command `maintenance:update:htaccess`
but: occ doesn't feature the command referenced to....... -.-
error =>
> [Symfony\Component\Console\Exception\CommandNotFoundException]
> There are no commands defined in the "maintenance:update" namespace.
> Did you mean one of these?
> maintenance:mimetype
> maintenance

also i tried to add `'front_controller_active' => true,` to the config despite only place this param gets mentioned is right here...

tl;dr
tried all the fixes in here and
...index.php/apps/files/ works as well as .../apps/files/
but every link within owncloud still directs to ...index.php/apps/files/

is this behavior right as it should be of today or did i miss something?

**Comment by Xenopathic** [comment]
@pandorra2 That command, and the config.php option, are only available in 9.0.3

**Comment by eppfel** [comment]
@pandorra2 @thejustsoul I found that `mod_env` was restricted to a whitelist, so  'front_controller_active' was not set, although `mod_rewrite` was active.
I was able to get it working with fiddling in the source code. More here #24972

**Comment by PVince81** [comment]
With all the information here, I'm confused again how to setup 9.0.2 to remove index.php and also 9.0.3.

Can someone provide step by step instructions on how to setup 9.0.2 (without front_controller_active) and 9.0.3 to make it work, also in update cases ? This is important for update testing and to make packaging properly compatible (see https://github.com/owncloud/core/issues/23512)

Thanks.

**Comment by PVince81** [comment]
So, I did some tests and here is the result: https://github.com/owncloud/documentation/issues/2205#issuecomment-229117430

It seems that having a writable ".htaccess" with properly configured web server for ".htaccess" seems to be enough. Installing or upgrading ownCloud would maintain the proper values there.

Next step: figure out what, how and why to use "front_controller_active" and also add info in the doc ticket.

**Comment by PVince81** [comment]
Ok, turns out there are some docs here: https://github.com/owncloud/documentation/issues/2431

**Comment by basert** [comment]
After enabling this, everything works fine except ajax "drap&drop" uploads. The POST request calls `<DOMAIN>/apps/files/ajax/upload.php`, the error log contains
`2016/07/02 17:34:35 [error] 23529#23529: *282 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught Error: Class 'OC' not found in /var/www/owncloud/apps/files/ajax/upload.php:36
Stack trace:
#0 {main}
  thrown in /var/www/owncloud/apps/files/ajax/upload.php on line 36" while reading response header from upstream, client: IP, server: HOST, request: "POST /apps/files/ajax/upload.php HTTP/1.1", upstream: "fastcgi://unix:/run/php/php7.0-fpm.sock:", host: "HOST"`

When I disable `fastcgi_param front_controller_active true` everything works as expected and the file gets uploaded.

**Comment by josh4trunks** [comment]
@basert upload.php shouldn't be one of the scripts owncloud executes. Are you using nginx? If so it sounds like your config isn't correct. Can you post it so I can take a look.

**Comment by basert** [comment]
Sure, thanks, here's a GIST of my config:
https://gist.github.com/basert/ca9dde89f3c5a54c604c25afc80f7cea

**Comment by RealRancor** [comment]
@basert You shouldn't use such an older Pull Request for such questions / issues. Either create a new issue if you think there is a bug within this feature or ask for support at either the users malinglist or the forums (see https://owncloud.org/support).

But just had a short look at your nginx config which doesn't match the config provided here:

https://doc.owncloud.org/server/9.0/admin_manual/installation/nginx_owncloud_9x.html#owncloud-in-the-webroot-of-nginx

**Comment by Delagen** [comment]
+1 for config for apache
